File: backend/templates/fibonacci/fibonacci.rb
#!/usr/bin/env ruby

# 1. 再帰版（メモ化付き）
class FibonacciRecursive
  def initialize
    @memo = {}
  end
  
  def calculate(n)
    return n if n <= 1
    @memo[n] ||= calculate(n - 1) + calculate(n - 2)
  end
end

# 2. イテレータ版
class FibonacciIterator
  include Enumerable
  
  def initialize(limit = Float::INFINITY)
    @limit = limit
  end
  
  def each
    a, b = 0, 1
    count = 0
    
    while count < @limit
      yield a
      a, b = b, a + b
      count += 1
    end
  end
end

# 3. 行列累乗版（高速計算）
class FibonacciMatrix
  def calculate(n)
    return n if n <= 1
    
    matrix = [[1, 1], [1, 0]]
    result = matrix_power(matrix, n)
    result[0][1]
  end
  
  private
  
  def matrix_multiply(a, b)
    [
      [a[0][0] * b[0][0] + a[0][1] * b[1][0], a[0][0] * b[0][1] + a[0][1] * b[1][1]],
      [a[1][0] * b[0][0] + a[1][1] * b[1][0], a[1][0] * b[0][1] + a[1][1] * b[1][1]]
    ]
  end
  
  def matrix_power(matrix, n)
    return [[1, 0], [0, 1]] if n == 0
    return matrix if n == 1
    
    if n.even?
      half = matrix_power(matrix, n / 2)
      matrix_multiply(half, half)
    else
      matrix_multiply(matrix, matrix_power(matrix, n - 1))
    end
  end
end

# 4. ジェネレータ版
def fibonacci_generator
  Enumerator.new do |y|
    a, b = 0, 1
    loop do
      y << a
      a, b = b, a + b
    end
  end
end

# 5. 黄金比を使った近似計算
class FibonacciGoldenRatio
  PHI = (1 + Math.sqrt(5)) / 2
  
  def calculate(n)
    ((PHI ** n - (-PHI) ** (-n)) / Math.sqrt(5)).round
  end
end

# デモンストレーション
puts "=== フィボナッチ数列の様々な実装 ==="
puts

# 1. 再帰版のデモ
puts "1. 再帰版（メモ化付き）"
fib_recursive = FibonacciRecursive.new
print "最初の10項: "
10.times { |i| print "#{fib_recursive.calculate(i)} " }
puts "\n"

# 2. イテレータ版のデモ
puts "2. イテレータ版"
print "最初の10項: "
FibonacciIterator.new(10).each { |n| print "#{n} " }
puts "\n"

# 3. 行列累乗版のデモ
puts "3. 行列累乗版（大きな数も高速）"
fib_matrix = FibonacciMatrix.new
print "n=50の値: #{fib_matrix.calculate(50)}"
puts "\n"

# 4. ジェネレータ版のデモ
puts "4. ジェネレータ版（遅延評価）"
print "100未満の項: "
fibonacci_generator.take_while { |n| n < 100 }.each { |n| print "#{n} " }
puts "\n"

# 5. 黄金比版のデモ
puts "5. 黄金比を使った近似計算"
fib_golden = FibonacciGoldenRatio.new
print "最初の10項: "
10.times { |i| print "#{fib_golden.calculate(i)} " }
puts "\n"

# おまけ：フィボナッチ数列の性質を調べる
puts "\n=== フィボナッチ数列の面白い性質 ==="

# 隣接項の比が黄金比に収束
puts "隣接項の比（黄金比#{FibonacciGoldenRatio::PHI}に収束）:"
fib = FibonacciIterator.new(15).to_a
(5..10).each do |i|
  ratio = fib[i].to_f / fib[i-1]
  puts "F(#{i})/F(#{i-1}) = #{ratio.round(6)}"
end

# 最初のn項の和
puts "\n最初のn項の和の性質（F(n+2) - 1に等しい）:"
(5..8).each do |n|
  sum = fib[0...n].sum
  puts "最初の#{n}項の和: #{sum} = F(#{n+2}) - 1 = #{fib[n+1]} - 1"
end

File: backend/templates/fibonacci/fibonacci.cs
using System;
using System.Collections.Generic;
using System.Linq;

// 1. 再帰版（メモ化付き）
public class FibonacciRecursive 
{
    private Dictionary<int, long> memo = new Dictionary<int, long>();
    
    public long Calculate(int n) 
    {
        if (n <= 1) return n;
        if (memo.ContainsKey(n)) return memo[n];
        memo[n] = Calculate(n - 1) + Calculate(n - 2);
        return memo[n];
    }
}

// 2. イテレータ版
public class FibonacciIterator 
{
    private int limit;
    
    public FibonacciIterator(int limit) 
    {
        this.limit = limit;
    }
    
    public IEnumerable<long> Generate() 
    {
        long a = 0, b = 1;
        int count = 0;
        
        while (count < limit) 
        {
            yield return a;
            long temp = a;
            a = b;
            b = temp + b;
            count++;
        }
    }
    
    public List<long> ToList() 
    {
        return Generate().ToList();
    }
}

// 3. 行列累乗版
public class Matrix 
{
    public long[,] M { get; set; }
    
    public Matrix(long a, long b, long c, long d) 
    {
        M = new long[2, 2] { { a, b }, { c, d } };
    }
    
    public static Matrix operator *(Matrix a, Matrix b) 
    {
        return new Matrix(
            a.M[0, 0] * b.M[0, 0] + a.M[0, 1] * b.M[1, 0],
            a.M[0, 0] * b.M[0, 1] + a.M[0, 1] * b.M[1, 1],
            a.M[1, 0] * b.M[0, 0] + a.M[1, 1] * b.M[1, 0],
            a.M[1, 0] * b.M[0, 1] + a.M[1, 1] * b.M[1, 1]
        );
    }
}

public class FibonacciMatrix 
{
    public long Calculate(int n) 
    {
        if (n <= 1) return n;
        var baseMatrix = new Matrix(1, 1, 1, 0);
        var result = MatrixPower(baseMatrix, n);
        return result.M[0, 1];
    }
    
    private Matrix MatrixPower(Matrix baseMatrix, int n) 
    {
        if (n == 0) return new Matrix(1, 0, 0, 1); // 単位行列
        if (n == 1) return baseMatrix;
        
        if (n % 2 == 0) 
        {
            var half = MatrixPower(baseMatrix, n / 2);
            return half * half;
        } 
        else 
        {
            return baseMatrix * MatrixPower(baseMatrix, n - 1);
        }
    }
}

// 4. 黄金比を使った近似計算
public class FibonacciGoldenRatio 
{
    public static readonly double PHI = (1.0 + Math.Sqrt(5.0)) / 2.0;
    
    public long Calculate(int n) 
    {
        return (long)Math.Round((Math.Pow(PHI, n) - Math.Pow(-PHI, -n)) / Math.Sqrt(5.0));
    }
}

// 5. LINQ版
public static class FibonacciLinq 
{
    public static IEnumerable<long> Generate(int count) 
    {
        return Enumerable.Range(0, count)
            .Aggregate(
                new List<long>(),
                (acc, i) => 
                {
                    if (i <= 1) acc.Add(i);
                    else acc.Add(acc[i - 1] + acc[i - 2]);
                    return acc;
                }
            );
    }
}

public class Program 
{
    public static void Main() 
    {
        Console.WriteLine("=== フィボナッチ数列の様々な実装 ===\n");
        
        // 1. 再帰版のデモ
        Console.WriteLine("1. 再帰版（メモ化付き）");
        var fibRecursive = new FibonacciRecursive();
        Console.Write("最初の10項: ");
        for (int i = 0; i < 10; i++) 
        {
            Console.Write($"{fibRecursive.Calculate(i)} ");
        }
        Console.WriteLine("\n");
        
        // 2. イテレータ版のデモ
        Console.WriteLine("2. イテレータ版");
        var fibIter = new FibonacciIterator(10);
        Console.Write("最初の10項: ");
        foreach (var value in fibIter.Generate()) 
        {
            Console.Write($"{value} ");
        }
        Console.WriteLine("\n");
        
        // 3. 行列累乗版のデモ
        Console.WriteLine("3. 行列累乗版（大きな数も高速）");
        var fibMatrix = new FibonacciMatrix();
        Console.WriteLine($"n=50の値: {fibMatrix.Calculate(50)}\n");
        
        // 4. 黄金比版のデモ
        Console.WriteLine("4. 黄金比を使った近似計算");
        var fibGolden = new FibonacciGoldenRatio();
        Console.Write("最初の10項: ");
        for (int i = 0; i < 10; i++) 
        {
            Console.Write($"{fibGolden.Calculate(i)} ");
        }
        Console.WriteLine("\n");
        
        // 5. LINQ版のデモ
        Console.WriteLine("5. LINQ版");
        var fibLinq = FibonacciLinq.Generate(15).ToList();
        Console.Write("最初の15項: ");
        Console.WriteLine(string.Join(" ", fibLinq));
        Console.WriteLine();
        
        // おまけ：フィボナッチ数列の性質を調べる
        Console.WriteLine("=== フィボナッチ数列の面白い性質 ===");
        
        // 隣接項の比が黄金比に収束
        Console.WriteLine($"隣接項の比（黄金比{FibonacciGoldenRatio.PHI:F6}に収束）:");
        var sequence = new FibonacciIterator(15).ToList();
        for (int i = 5; i <= 10; i++) 
        {
            double ratio = (double)sequence[i] / sequence[i - 1];
            Console.WriteLine($"F({i})/F({i - 1}) = {ratio:F6}");
        }
        
        // 最初のn項の和
        Console.WriteLine("\n最初のn項の和の性質（F(n+2) - 1に等しい）:");
        for (int n = 5; n <= 8; n++) 
        {
            long sum = sequence.Take(n).Sum();
            Console.WriteLine($"最初の{n}項の和: {sum} = F({n + 2}) - 1 = {sequence[n + 1]} - 1");
        }
        
        // C#の特殊機能を使った実装
        Console.WriteLine("\n=== C#の特殊機能を使った実装 ===");
        
        // Funcデリゲート版
        Func<int, long> fibFunc = null;
        var memo = new Dictionary<int, long>();
        fibFunc = n => n <= 1 ? n : 
            memo.ContainsKey(n) ? memo[n] : 
            memo[n] = fibFunc(n - 1) + fibFunc(n - 2);
        
        Console.WriteLine($"Funcデリゲート版 F(10): {fibFunc(10)}");
        
        // ローカル関数版
        long FibLocal(int n) 
        {
            static Dictionary<int, long> localMemo = new Dictionary<int, long>();
            if (n <= 1) return n;
            if (localMemo.ContainsKey(n)) return localMemo[n];
            return localMemo[n] = FibLocal(n - 1) + FibLocal(n - 2);
        }
        
        Console.WriteLine($"ローカル関数版 F(10): {FibLocal(10)}");
    }
}